src/coreclr/src/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs

```diff
@@ -3873,43 +3867,71 @@ private void CreateInstanceCheckThis()
-            internal MethodAttributes _ctorAttributes;
-            internal CtorDelegate? _ctor;
-
-            // Lazy initialization was performed
-            internal volatile bool _isFullyInitialized;
-
-            private static ConstructorInfo? s_delegateCtorInfo;
+            internal volatile delegate*<MethodTable*, object?> _pfnNewobj;
+            internal readonly delegate*<object?, void> _pfnCtor;
+            internal MethodTable* _pMT;
+            internal readonly bool _ctorIsPublic;
 
             internal ActivatorCache(RuntimeMethodHandleInternal rmh)
             {
-                _hCtorMethodHandle = rmh;
+                if (rmh.IsNullHandle())
+                {
+                    // Value type with no explicit constructor, which means the "default" ctor
+                    // is implicitly public and no-ops.
+
+                    _ctorIsPublic = true;
+                    _pfnCtor = &GC.KeepAlive; // a convenient no-op fn
+                }
+                else
+                {
+                    // Reference type with a parameterless ctor.
+
+                    _ctorIsPublic = (RuntimeMethodHandle.GetAttributes(rmh) & MethodAttributes.Public) != 0;
+                    _pfnCtor = (delegate*<object?, void>)RuntimeMethodHandle.GetFunctionPointer(rmh);
+                }
+
+                Debug.Assert(_pfnCtor != null);
             }
 
-            private void Initialize()
+            private void Initialize(RuntimeType type)
             {
-                if (!_hCtorMethodHandle.IsNullHandle())
-                {
-                    _ctorAttributes = RuntimeMethodHandle.GetAttributes(_hCtorMethodHandle);
+                Debug.Assert(type != null);
 
-                    // The default ctor path is optimized for reference types only
-                    ConstructorInfo delegateCtorInfo = s_delegateCtorInfo ??= typeof(CtorDelegate).GetConstructor(new Type[] { typeof(object), typeof(IntPtr) })!;
+                // If we reached this point, we already know that the construction information
+                // can be cached, so all of the runtime reflection checks succeeded. We just
+                // need to special-case Nullable<T> (to return null).
+                //
+                // No synchronization is needed in this method since we have marked the _pfnNewobj
+                // field as volatile, and if there's multi-threaded access all threads will agree
+                // on the values to write anyway.
+                //
+                // !! IMPORTANT !!
+                // Don't assign the function pointer return value of GetNewobjHelperFnPtr directly
+                // to the backing field, as setting the field marks initialization as complete.
+                // Be sure to perform any last-minute checks *before* setting the backing field.
 
-                    // No synchronization needed here. In the worst case we create extra garbage
```

Comment:
The cache initialization was expensive before this change due to this MethodInfo.Invoke call. It is why we have tried hard to delay the initialization to the second call. This change fixes this problem and this scheme does not make sense anymore.

I think we should:
- Delete the `bool fillCache`. It should be fine to use the cache for all cases.
- Use the same codepath to allocate the first time through, ie change `RuntimeTypeHandle.CreateInstance` to return the constructor pointer, whether it is public, etc. This will eliminate problems with subtle behavior differences between the first time call and the subsequent calls.